src/casync-http.c

```diff
@@ -11,67 +12,1066 @@
+                cd = get_curl_private(handle);
+                if (cd)
+                        chunk_data_free(cd);
+
+                curl_easy_cleanup(handle);
+        }
+
+        while ((handle = queue_pop(dl->completed))) {
+                cd = get_curl_private(handle);
+                if (cd)
+                        chunk_data_free(cd);
+
+                curl_easy_cleanup(handle);
+        }
+
+        free(dl->store_url);
+        queue_free(dl->ready);
+        queue_free(dl->inprogress);
+        queue_free(dl->completed);
+        curl_multi_cleanup(dl->multi);
+        ca_remote_unref(dl->remote);
+
+        free(dl);
+}
+
+static inline void ca_chunk_downloader_freep(CaChunkDownloader **dl) {
+        ca_chunk_downloader_free(*dl);
+}
+
+static CaChunkDownloader *ca_chunk_downloader_new(CaRemote *rr, const char *store_url) {
+        CaChunkDownloader *dl = NULL;
+        int r, i;
+
+        dl = new0(CaChunkDownloader, 1);
+        if (!dl)
+                goto fail;
+
+        dl->remote = ca_remote_ref(rr);
+
+        r = make_curl_multi_handle(&dl->multi);
+        if (r < 0)
+                goto fail;
+
+        dl->ready = queue_new();
+        if (!dl->ready)
+                goto fail;
+
+        dl->inprogress = queue_new();
+        if (!dl->inprogress)
+                goto fail;
+
+        dl->completed = queue_new();
+        if (!dl->completed)
+                goto fail;
+
+        for (i = 0; i < arg_max_active_chunks; i++) {
+                CURL *handle = NULL;
+                ChunkData *cd = NULL;
+
+                cd = chunk_data_new();
+                if (!cd)
+                        goto fail;
+
+                r = make_curl_easy_handle(&handle, write_chunk, &cd->buffer, cd);
+                if (r < 0)
+                        goto fail;
+
+                queue_push(dl->ready, handle);
+        }
+
+        dl->store_url = strdup(store_url);
+        if (!dl->store_url)
+                goto fail;
+
+        return dl;
+
+fail:
+        ca_chunk_downloader_free(dl);
+        return NULL;
+}
+
+#define AVERAGE(sum, n) (n > 0 ? sum / n : sum)
+
+static void ca_chunk_downloader_display_stats(CaChunkDownloader *dl) {
+        if (!arg_verbose)
+                return;
+
+        log_info("--- Chunk Downloader Stats ---");
+        log_info("Iterations: %" PRIu64, dl->n_iterations);
+        log_info("CURL handles totals: added=%" PRIu64 ", removed=%" PRIu64,
+                 dl->inprogress->n_added, dl->inprogress->n_removed);
+        log_info("Chunks: %" PRIu64 " put to remote", dl->completed->n_removed);
+        log_info("Queues average size: inprogress=%" PRIu64 ", completed=%" PRIu64,
+                 AVERAGE(dl->sum_inprogress_len, dl->n_iterations),
+                 AVERAGE(dl->sum_completed_len, dl->n_iterations));
+        log_info("Queues current size: inprogress=%" PRIu64 ", completed=%" PRIu64,
+                 dl->inprogress->len, dl->completed->len);
```

Comment:
i think this would be more readable in a single log_info() expression with a string with embedded newlines

Reply:
Done. Actually these logs are useful as long as I'm working and re-working all that, but when things are settled, I think it's best to make it `log_debug`, as it's not useful for user at all. We could also even comment it out, at your preference, doesn't matter.